doc/aaa/TACACS+ Design.md

```diff
@@ -0,0 +1,521 @@
+
+## 4.5 CLI
+ - The existing TACACS+ server config command will not change.
+ - Add following command to enable/disable TACACS+ authorizarion.
+```
+    // authorization with TACACS+ server and local
+    config aaa authorization tacacs local
+    
+    // authorization with TACACS+ server
+    config aaa authorization tacacs
+    
+    // authorization with TACACS+ local
+    config aaa authorization local
+```
+
+ - Add following command to enable/disable TACACS+ accounting.
+```
+    // accounting with TACACS+ server and local syslog
+    config aaa accounting tacacs local
+
+    // accounting with TACACS+ server
+    config aaa accounting tacacs
+
+    // accounting with local syslog
+    config aaa accounting local
+```
+
+ - When config AAA authorization with "no" prefix, SONiC will use local authorization, so following commands have same effect
+```
+    no config aaa authorization tacacs local
+    no config aaa authorization local
+    config aaa authorization local
+```
+
+ - When config AAA accounting with "no" prefix, SONiC will use stop accounting, following command have same effect.
+```
+    no config aaa authorization tacacs local
+    no config aaa authorization tacacs
+    no config aaa authorization local
+```
+# 5 Error handling
+ - Bash plugin for authorization will return error code [Bash](#bash). and patched Bash will:
+   - Output error log to syslog.
+   - Output error message to stdout.
+ - [audisp-tacplus](#audisp-tacplus) will return errors as per [Auditd](#auditd)  respectively.
+
+# 6 Serviceability and Debug
+ - The Bash plugin and [audisp-tacplus](#audisp-tacplus) can be debugged by enabling the debug
+field of the AAA|authentication key. (Please see ConfigDB AAA Table
+Schema in [TACACS+ Authentication](#TACPLUS-Authentication)). 
+
+# 7 Unit Test
+## 7.1 Unit test for source code
+ - All patch code in Bash and Bash plugin should have 100% code coverage.
+ - Bash plugin test, all TACACS+ server not reachable test:
+```
+    Verify TACACS+ authorization failed.
+```
+
+ - Bash plugin test, partial TACACS+ server accessable, and user command config as allowed on all server.
+```
+    Verify TACACS+ authorization passed.
+```
+
+ - Bash plugin test, partial TACACS+ server accessable, and user command config as reject on all server.
+```
+    Verify TACACS+ authorization rejected.
+```
+
+ - Bash plugin test, partial TACACS+ server accessable, and user command config as reject on accessable server, and allow on not accessable server.
+```
+    Verify TACACS+ authorization rejected.
+```
+
+ - Bash plugin test, partial TACACS+ server accessable, and user command config as allow on accessable server, and reject on not accessable server.
+```
+    Verify TACACS+ authorization passed.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, all TACACS+ server accessable.
+```
+    Verify TACACS+ accounting succeeded.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, all TACACS+ server not accessable.
```

Comment:
>all TACACS+ server not accessable [](http://example.com/codeflow?start=43&length=33)

Add a test case which tacacs is available for a while, user could login, run some comands, and suddently tacacs is not accessible, and test for the expected behavior.

Reply:
Test added:

config AAA authorization with TACACS+ only:

- when user login server are accessible.
- user run some command in whitelist and server are accessible.
- then all  server not accessible, and run some command